librato_python_web/instrumentor/web/cherrypy_.py

```diff
@@ -39,55 +39,51 @@
+        # call the request function
+        response = func(*args, **keywords)
 
-            if response.status:
-                telemetry.count('web.status.%sxx' % response.status[0:1])
-            return response
-        except Exception as e:
-            telemetry.count('web.errors')
-            raise e
-        finally:
-            try:
-                elapsed, net_elapsed = Timing.pop_timer()
-                telemetry.record('web.response.latency', elapsed)
-                telemetry.record('app.response.latency', net_elapsed)
-            except:
-                logger.exception('Teardown handler failed')
-                raise
-    return decorator
+        if response.status:
+            telemetry.count('web.status.%sxx' % response.status[0:1])
+        return response
+    except Exception as e:
+        telemetry.count('web.errors')
+        raise e
+    finally:
+        try:
+            elapsed, net_elapsed = Timing.pop_timer()
+            telemetry.record('web.response.latency', elapsed)
+            telemetry.record('app.response.latency', net_elapsed)
+        except:
+            logger.exception('Teardown handler failed')
+            raise
 
 
-def _cherrypy_wsgi_call(f):
-    def decorator(*args, **keywords):
-        t = time.time()
-        try:
-            return f(*args, **keywords)
-        finally:
-            elapsed = time.time() - t
-            telemetry.record('wsgi.response.latency', elapsed)
-    return decorator
+def _cherrypy_wsgi_call(func, *args, **keywords):
+    t = time.time()
+    try:
+        return func(*args, **keywords)
+    finally:
+        elapsed = time.time() - t
+        telemetry.record('wsgi.response.latency', elapsed)
 
 
 class CherryPyInstrumentor(BaseInstrumentor):
     modules = {'cherrypy._cptree': ['Application'], 'cherrypy._cprequest': ['Request']}
 
     def __init__(self):
-        super(CherryPyInstrumentor, self).__init__(
-            {
-                'cherrypy._cptree.Application.__call__': function_wrapper_factory(_cherrypy_wsgi_call, enable_if=None,
-                                                                                  state='web'),
-                'cherrypy._cprequest.Request.run': function_wrapper_factory(_cherrypy_respond_wrapper, enable_if=None,
-                                                                            state='wsgi'),
-            }
-        )
+        super(CherryPyInstrumentor, self).__init__()
 
     def run(self):
+        self.set_wrapped(
+            {
+                'cherrypy._cptree.Application.__call__': get_conditional_wrapper(_cherrypy_wsgi_call, enable_if=None,
+                                                                                 state='web'),
+                'cherrypy._cprequest.Request.run': get_conditional_wrapper(_cherrypy_respond_wrapper, enable_if=None,
+                                                                           state='wsgi'),
+            })
```

Comment:
The only real work the conditional wrapper does here is to push the state ('web' or 'wsgi') onto the stack. It also guards against possible (but very unlikely) recursive calls to the instrumented functions.
I'll target this as something to refine as we go along.